### ISSUE TITLE
`#eval` command with unresolved type metavariables.

### ISSUE THREAD

**Issue by leodemoura** [leanprover/lean4#1256]
The `#eval` command currently produces nasty error messages when the given term contains unresolved metavariables.
For example, in David's book, he uses the following example
```lean
def List.head? {α : Type} (xs : List α) : Option α :=
  match xs with
    | [] => none
    | y :: _ => some y

#eval [].head?
```
and has to explain the error message
```
expression
  _root_.List.head? []
has type
  Option ?m.13533
but instance
  Lean.MetaEval (Option ?m.13533)
failed to be synthesized, this instance instructs Lean on how to display the resulting value, recall that any type implementing the `Repr` class also implements the `Lean.MetaEval` class
```
This looks very cryptic to new users. We have talked about this issue in the past, but we never implemented a solution for it.
Should we just assign unassigned metavariables `?m : Type u` to `PUnit.punit.{u}` in the `#eval` command?
Any other ideas? Sorry if I forgot them.

@kha @gebner @EdAyers @Vtec234

**Comment by EdAyers** [comment]
I think it's a good idea.

I encountered this while writing some lenses: I had `fst : Lens α β (α × χ) (β × χ)` and `view : Lens α β σ τ → σ → α`. So `#eval view fst ("hello", "world")` fails since `β` is unassigned. Is this a problem outside of `#eval` too?

**Comment by leodemoura** [comment]
> Is this a problem outside of #eval too?

Which other commands do you have in mind? `#reduce`? For `#reduce t`, we can still try to reduce `t` even if it contains unassigned metavariables, and the result will be useful.
For declarations, I think it is better to throw an error if there are unassigned metavariables `?m : Type u`.

**Comment by Kha** [comment]
For self-documentation purposes we might want to use an alias such as `abbrev Unused := PUnit`

**Comment by gebner** [comment]
> This looks very cryptic to new users.

Indeed, the error message could be better.  If you wrote it as `def eval := [].head?`, you'd get "don't know how to synthesize implicit argument", which is at least clear about what's wrong.


Personally I find the automatic unit-defaulting a bit too magic though:
```lean
#eval decide (∀ x y, x = y) -- true?!
```

It seems reasonable to me that you'd get the same error message whether you write `#eval` or `def`.  Users need to learn about the error message either way, so we're not helping them by defaulting it in one place only.  IMHO the question here is whether unit-defaulting is more *ergonomic*.  I don't think I've ever ran into unassigned metavariables myself with `#eval`.  Missing `Repr` instances, yes.  Missing `MetaEval` instances (`CommandElabM`), yes.  Annoying requirement to specify monad (`show TermElabM _ from do`), yes.  But at least in my experience, the unassigned metavariables are very rare.

**Comment by leodemoura** [comment]
@gebner 

> Personally I find the automatic unit-defaulting a bit too magic though:
>
> #eval decide (∀ x y, x = y) -- true?!

Good point.

> It seems reasonable to me that you'd get the same error message whether you write #eval or def. Users need to learn about the error message either way, so we're not helping them by defaulting it in one place only.

Good. I will do this.

> IMHO the question here is whether unit-defaulting is more ergonomic. I don't think I've ever ran into unassigned metavariables myself with #eval. Missing Repr instances, yes. Missing MetaEval instances (CommandElabM), yes. Annoying requirement to specify monad (show TermElabM _ from do), yes. But at least in my experience, the unassigned metavariables are very rare.

I agree these issues are all annoying. 

> Missing Repr instances, yes.

We can try to synthesize the missing `Repr`s on the fly using `deriving Repr`, but `deriving Repr` can still fail. Perhaps, the best solution here is to just improve the error message. What do you think?

> Missing MetaEval instances (CommandElabM), yes.

I will add this one.

> Annoying requirement to specify monad (show TermElabM _ from do), yes.

The `do` notation currently throws an error if the `Monad` is not available. We could relax this restriction on the `do` notation, most of the time, users want to use `#eval do ...` on very simple code blocks, and auto-lifting is probably not needed. It also helps that we don't have the `Id` feature anymore.

**Comment by Kha** [comment]
> We can try to synthesize the missing `Repr`s on the fly using `deriving Repr`

A bit of an aside, but that would be incredibly useful for `dbg_trace`

**Comment by leodemoura** [comment]
> Personally I find the automatic unit-defaulting a bit too magic though:
> #eval decide (∀ x y, x = y) -- true?!

@gebner What about assigning `PUnit.{u}` only to metavariables that occur in the resulting type. Then, the following examples 
would work 
```lean
#eval none 
#eval [] 
```
but 
```lean
#eval decide (∀ x y, x = y)
```
would still fail as expected.

**Comment by gebner** [comment]
Then it would be mysterious why `#eval [].head?` works, but `#eval [].length` fails.  We could do the defaulting after type-class synthesis though, then the result shouldn't depend on the defaulted type.

But I'm not feeling very strongly about this.  We already use defaulting for `Nat`.  However the Nat-defaulting works everywhere, whereas the unit-defaulting would be a special adhoc feature that only works in a single place, adding complexity to the language.

**Comment by leodemoura** [comment]
I pushed a few improvements for the `#eval` command. The only feature missing from the suggestions above is the on-demand `Repr` synthesis. The key problem is that we currently don't have a mechanism for explaining TC resolution failures. This is doable, but I will not work on this feature right now. 
I am closing this issue since the original issue has been addressed. The command `#eval [].head?` now produces the same  error produced by `def aux := [].head?`.
